tests/functional/test_ec2.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from tests import unittest
+from botocore.stub import Stubber, StubAssertionError, ANY
+import botocore.session
+
+
+class TestIdempotencyToken(unittest.TestCase):
+    def setUp(self):
+        self.function_name = 'purchase_scheduled_instances'
+        self.region = 'us-west-2'
+        self.session = botocore.session.get_session()
+        self.client = self.session.create_client(
+            'ec2', self.region)
+        self.stubber = Stubber(self.client)
+        self.service_response = {}
+        self.params_seen = []
+
+        # Record all the parameters that get seen
+        self.client.meta.events.register_first(
+            'before-call.*.*',
+            lambda model, params, **kwargs: self.params_seen.extend(
+                params['body'].keys()),
+            unique_id='TestIdempotencyToken')
+
+    def test_provided_idempotency_token(self):
+        expected_params = {
+            'PurchaseRequests': [
+                {'PurchaseToken': 'foo',
+                 'InstanceCount': 123}],
+            'ClientToken': ANY
+        }
+        self.stubber.add_response(
+            self.function_name, self.service_response, expected_params)
+
+        with self.stubber:
+            try:
```

Comment:
I do not think you need this try, except, finally for a couple reasons:

1) The ``self.params_seen`` will always get reset to an empty list because it is defined in ``setUp`` and ``setUp`` gets called for each test.
2) There does not appear to be too much benefit for catching the StubAssertionError and then re-raising it. It seems unneeded.

Let me know what you think.